Stm32 smps support

manoj153 opened the pull request:
- Config power to use SMPS when enabled

manoj153 committed:
drivers: clock_control: stm32h7: Add logic to handle SMPS config

Some STM32 SoC supports an internal SMPS

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 committed:
soc: arm: st_stm32: add kconfig entry for STM32 SMPS

Add support for SMPS

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

erwango commented:
@ABOSTM, Does it fix https://github.com/zephyrproject-rtos/zephyr/issues/34732 ?

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -260,7 +260,11 @@ static uint32_t get_hclk_frequency(void)
 static int32_t prepare_regulator_voltage_scale(void)
 {
 	/* Make sure to put the CPU in highest Voltage scale during clock configuration */
+#ifdef CONFIG_POWER_SUPPLY_SMPS
```
Like you did in soc_m7.c:
you should check SMPS flag (the one from CMSIS)

ABOSTM commented on drivers/clock_control/clock_stm32_ll_h7.c:
```diff
@@ -276,7 +280,11 @@ static int32_t optimize_regulator_voltage_scale(uint32_t sysclk_freq)
 	/* LL_PWR_REGULATOR_SCALE3 is lowest power consumption */
 	/* Must be done in accordance to the Maximum allowed frequency vs VOS*/
 	/* See RM0433 page 352 for more details */
+#ifdef CONFIG_POWER_SUPPLY_SMPS
```
dito

ABOSTM commented on soc/arm/st_stm32/stm32h7/soc_m7.c:
```diff
@@ -80,7 +80,7 @@ static int stm32h7_init(const struct device *arg)
 
 	/* Power Configuration */
-#ifdef SMPS
+#if defined(SMPS) && defined(CONFIG_POWER_SUPPLY_SMPS)
```
Could be interesting, in case CONFIG_POWER_SUPPLY_SMPS is enabled and SMPS is not, to raise a compilation error.
(CONFIG_POWER_SUPPLY_SMPS  should not be set when SMPS is not supported)

ABOSTM commented on soc/arm/st_stm32/common/Kconfig.soc:
```diff
@@ -21,3 +21,8 @@ config USE_STM32_ASSERT
 	bool "STM32Cube HAL and LL drivers asserts"
 	help
 	  Enable asserts in STM32Cube HAL and LL drivers.
+
+config POWER_SUPPLY_SMPS
+	bool "Enable STM32 internal SMPS"
+	help
+	  Enable SMPS Power Supply.
```
This config is currently supported only for STM32H7 series.
Can you encapsulate it with :
```
if SOC_SERIES_STM32H7X
...
endif 
```

ABOSTM commented:
> @ABOSTM, Does it fix #34732 ?

This is for sure a good step, but we still need to define this switch in board stm32h747i_disco

ABOSTM commented on soc/arm/st_stm32/common/Kconfig.soc:
```diff
@@ -21,3 +21,12 @@ config USE_STM32_ASSERT
+
+if SOC_SERIES_STM32H7X
+
+config POWER_SUPPLY_SMPS
```
@erwango, other STM32 series have LDO/SMPS (even if not yet supported in Zephyr), for example stm32wl, stm32u5, ...
So it may be good to keep this swich in a common Kconfig.

manoj153 replied:
@erwango advice if what ABOSTM comments is rational else I will make the amendments

erwango replied:
ok, let's keep it here then.
But, please use more compact "depends on", when only one symbol is impacted (instead of "if FOO").

manoj153 replied:
Done

erwango replied:
Would you mind pushing this in the initial commit ?

manoj153 replied:
Done

manoj153 committed:
soc: arm: st_stm32: use SMPS power supply only if enabled

Use SMPS power supply only if enabled.

The default power supply configuration for the
NUCLEO board with -Q subfix is SMPS,
so it's essential to match with hardware configuration
to avoid deadlocks due to mismatch.

if a custom board with LDO configuration is in use,
then no need to enable `CONFIG_POWER_SUPPLY_SMPS`

Signed-off-by: Manojkumar Subramaniam <manoj@electrolance.com>

manoj153 commented:
@ABOSTM is all good now? If yes +1 this

ABOSTM reviewed (approved):
LGTM
